pkg/scheduler/eventhandlers.go

```diff
@@ -33,6 +33,67 @@ import (
 	"k8s.io/kubernetes/pkg/scheduler/internal/queue"
 )
 
+func parse(obj interface{}) *v1.Pod {
+	switch t := obj.(type) {
+	case *v1.Pod:
+		return t
+	case cache.DeletedFinalStateUnknown:
+		if pod, ok := t.Obj.(*v1.Pod); ok {
+			return pod
+		}
+		utilruntime.HandleError(fmt.Errorf("unable to convert object %T to *v1.Pod", obj))
+	default:
+		utilruntime.HandleError(fmt.Errorf("unable to handle object %T", obj))
+	}
+	return nil
+}
+
+func (sched *Scheduler) onPodAdd(obj interface{}) {
+	pod := parse(obj)
+	if pod == nil {
+		return
+	}
+	// TODO(Huang-Wei): take responsibleForPod() into consideration
+	if assignedPod(pod) {
+		sched.addPodToCache(obj)
+	} else {
+		sched.addPodToSchedulingQueue(obj)
+	}
+}
+
+func (sched *Scheduler) onPodUpdate(old, new interface{}) {
+	oldPod, newPod := parse(old), parse(new)
+	if oldPod == nil || newPod == nil {
+		return
+	}
+	// TODO(Huang-Wei): take responsibleForPod() into consideration
+	oldAssigned, newAssigned := assignedPod(oldPod), assignedPod(newPod)
+	if oldAssigned && newAssigned {
+		sched.updatePodInCache(oldPod, newPod)
+	} else if !oldAssigned && !newAssigned {
+		sched.updatePodInSchedulingQueue(oldPod, newPod)
+	} else if oldAssigned && !newAssigned {
+		sched.deletePodFromCache(oldPod)
+		sched.addPodToSchedulingQueue(newPod)
+	} else if !oldAssigned && newAssigned {
+		sched.deletePodFromSchedulingQueue(oldPod)
```

Comment:
We need to make sure the behaviors are equivalent. For example, if the scheduler name never matched this scheduler, then we don't need to call this function, which is potentially expensive.

Reply:
Yeap, I left a TODO there.